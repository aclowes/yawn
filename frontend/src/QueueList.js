import React from 'react';
import {Table, Alert, Button} from 'react-bootstrap';

import API from "./API";

export default class QueueList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {queues: null, error: null};
  }

  componentDidMount() {
    API.get(`/api/queues/`, (payload, error) => {
      this.setState({queues: payload, error});
    });
  }

  purgeQueue(id) {
    API.patch(`/api/queues/${id}/`, {purge: true}, (payload, error) => {
      if (error) {
        this.setState({error});
      } else {
        // insert the updated queue record
        this.setState({
          queues: this.state.queues.map((queue) => {
            return queue.id === payload.id ? payload : queue;
          })
        });
      }
    });
  }

  renderRows() {
    if (this.state.queues === null) {
      return (
        <tr>
          <td colSpan="3">Loading...</td>
        </tr>
      )
    } else {
      return this.state.queues.map((queue) => {
        const terminate = () => this.purgeQueue(queue.id);
        return (
          <tr key={queue.id}>
            <td>{queue.name}</td>
            <td>{queue.message_count}</td>
            <td>
              <Button bsSize="small" onClick={terminate}>
                Purge</Button>
            </td>
          </tr>
        )
      })
    }
  }

  render() {
    return (
      <div>
        <h3>Queues</h3>
        {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
        <Table striped bordered condensed hover>
          <thead>
          <tr>
            <th>Name</th>
            <th>Message Count</th>
            <th>Actions</th>
          </tr>
          </thead>
          <tbody>
          {this.renderRows()}
          </tbody>
        </Table>
      </div>
    )
  }
}
