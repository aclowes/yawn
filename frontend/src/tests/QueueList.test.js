import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import QueueList from '../QueueList';
import {mockAPI} from './mocks'

it('QueueList loading', () => {
  mockAPI([null]);
  const component = ReactTestRenderer.create(<QueueList />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('QueueList failure', () => {
  mockAPI();
  const component = ReactTestRenderer.create(<QueueList />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('QueueList success', () => {
  mockAPI([[{"id": 1, "message_count": 0, "name": "default"}]]);
  const component = ReactTestRenderer.create(<QueueList />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
