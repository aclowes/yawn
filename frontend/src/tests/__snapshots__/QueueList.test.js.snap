// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueList failure 1`] = `
<div>
  <h3>
    Queues
  </h3>
  <div
    className="alert alert-danger"
    role="alert"
  >
    some error message
  </div>
  <table
    className="table table-striped table-bordered table-condensed table-hover"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Message Count
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan="3"
        >
          Loading...
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`QueueList loading 1`] = `
<div>
  <h3>
    Queues
  </h3>
  <table
    className="table table-striped table-bordered table-condensed table-hover"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Message Count
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan="3"
        >
          Loading...
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`QueueList success 1`] = `
<div>
  <h3>
    Queues
  </h3>
  <table
    className="table table-striped table-bordered table-condensed table-hover"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Message Count
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          default
        </td>
        <td>
          0
        </td>
        <td>
          <button
            className="btn btn-sm btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Purge
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
